### PR TITLE
Download functionality on embedded dashboards (feature 3847)

### PR DESCRIPTION
**Pull request by r-kot**
Based on https://github.com/metabase/metabase/pull/6322
Requirments: https://github.com/metabase/metabase/pull/6322#issuecomment-353668283
Proposal: https://github.com/metabase/metabase/issues/3847

> - [x] I think this button should only appear when hovering on a dashboard card. It feels like too much visual noise to display it persistently on every card to me.
> - [x] There's a different arrow icon you can use called download. I know you were intending to match the current download icon we're using on question pages, but we'll be swapping that icon out for the download one there as well in a separate issue.

-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

**Comment by mazameli** [comment]
Hi, @r-kot. I'm not seeing the download button when hovering hover a dashboard card. Is this working for you locally?

**Comment by r-kot** [comment]
Hi, @mazameli. It works only for embedded dashboards ("Embed this dashboard in an application") with JWT token.

File: src/metabase/api/embed.clj [hunk 1]
```diff
@@ -417,4 +516,26 @@
...
+   [token export-format dashcard-id & query-params]
+    {export-format dataset-api/ExportFormat}
+      (dataset-api/as-format export-format
+      (let [cards-id (first (csv/read-csv  (query-params :card-ids)))
+            query-params (dissoc query-params :card-ids)
+            cards-data (into [] (map (fn [card-id] (card-for-signed-token token dashcard-id (Integer/parseInt card-id) query-params)) cards-id))
+            combined (reduce combine-cards-data cards-data)
+        ]
+        combined
+    )))
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/bbatsov/clojure-style-guide#gather-trailing-parens

File: src/metabase/api/embed.clj [hunk 2]
```diff
@@ -33,7 +33,8 @@
              [embed :as eu]
              [schema :as su]]
             [schema.core :as s]
-            [toucan.db :as db]))
+            [toucan.db :as db]
+            [clojure.data.csv :as csv]))
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#keep-requires-ordered

File: src/metabase/api/embed.clj [hunk 3]
```diff
@@ -70,9 +71,9 @@
   [object-embedding-params token-params user-params]
   ;; TODO - maybe make this log/debug once embedding is wrapped up
   (log/debug "Validating params for embedded object:\n"
-             "object embedding params:" object-embedding-params
-             "token params:"            token-params
-             "user params:"             user-params)
+            "object embedding params:" object-embedding-params
+            "token params:"            token-params
+            "user params:"             user-params)
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/bbatsov/clojure-style-guide#vertically-align-fn-args

File: src/metabase/api/embed.clj [hunk 4]
```diff
@@ -153,8 +154,8 @@
    `value`."
   [parameters parameter-values]
   (when (seq parameters)
-    (for [param parameters
-          :let  [value (get parameter-values (keyword (:slug param)))]
+  (for [param parameters
+        :let  [value (get parameter-values (keyword (:slug param)))]
           :when (some? value)]
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/bbatsov/clojure-style-guide#vertically-align-fn-args

File: src/metabase/api/embed.clj [hunk 5]
```diff
@@ -188,7 +189,98 @@
   not automatically converted. Thus we must do it ourselves here to make sure things are done as we'd expect."
   [query-params]
   (m/map-keys keyword query-params))
-
+(defn- mapify-row [row columns mergeKeysSet]
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/bbatsov/clojure-style-guide#empty-lines-between-top-level-forms

File: src/metabase/api/embed.clj [hunk 6]
```diff
@@ -188,7 +189,98 @@
...
+    (if (not ( = (count row) (count columns)))
+      (println "error ambiguous mapping col count not equal to row count" row columns)
+      (if (= len (count columns))
+        (assoc {} rowkey array)
+        (let [colValue (nth row index)
+              colName (nth columns index)]
+          (if (not (contains? mergeKeysSet colName))
+            (recur (inc len) (inc index) rowkey (conj array colValue))
+            (recur (inc len) (inc index) (str rowkey colValue) array)
+            ))))))
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/bbatsov/clojure-style-guide#gather-trailing-parens

File: src/metabase/api/embed.clj [hunk 7]
```diff
@@ -188,7 +189,98 @@
...
+(defn- merge-query-params [data1 data2]
+  {:json_query (data1 :json_query),
+   :status (data1 :status)} )
+
+(defn- merge-cols [data1 data2 mergeKeysSet allColumns]
+  (let [cols (concat (data1 :cols)
+                     (for [col (data2 :cols)
+                           :when (not (contains? mergeKeysSet (col :name)))]
+                       col)) ]
+    {:cols (into []  cols)}) )
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: https://github.com/bbatsov/clojure-style-guide#gather-trailing-parens

File: src/metabase/api/embed.clj [hunk 8]
```diff
@@ -188,7 +189,98 @@
...
+                                merge-columns]]
+      (let [out      (f data1 data2 mergeKey allColumns)
+            data (if
+                   (nil? out)        data (merge data out))]
+        (if-not (seq more)
+          (merge {:data data } (merge-query-params card1 card2))
+          (recur data more)))))))
+
+
+;;; ------------------------------------------------------------ Card Fns used by both /api/embed and /api/preview_embed ------------------------------------------------------------
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: What is the reason for this change?

File: src/metabase/api/embed.clj [hunk 9]
```diff
@@ -334,7 +427,9 @@
       :params   <parameters>}
 
    Additional dashboard parameters can be provided in the query string, but params in the JWT token take precedence."
-  [token dashcard-id card-id & query-params]
+  {:style/indent 1}
+  [token dashcard-id card-id query-params]
+  (println token dashcard-id card-id query-params)
```
Comment (camsaul) [inline src/metabase/api/embed.clj]: use one of the `log/` functions if you want to log something

**Comment by majdnmaz** [comment]
"Embeding dashboard in an application" with JWT token. where should I add the Insert this code snippet in your server code to generate the signed embedding URL. 
which file on my server.
 have installed composer require lcobucci/jwt

Please help

File: frontend/src/metabase/lib/urls.js [hunk 10]
```diff
@@ -34,6 +26,16 @@ export function question(cardId, hash = "", query = "") {
     : `/question${query}${hash}`;
 }
 
+export const extractQueryParams = (query: Object): Array => {
+  return Object.entries(query).flatMap(flattenParam);
```
Comment (tlrobinson) [inline frontend/src/metabase/lib/urls.js]: I believe `flatMap` isn't implemented in a number of browsers. Something like this should work isntead

```
  return [].concat(...Object.entries(query).map(flattenParam));
```

**Review by tlrobinson — COMMENTED** [review]
@metabase/core-developers Do we want to support this in private and public dashboards as well, for completeness?

**Comment by tlrobinson** [comment]
The button color is a bit dark, we probably want to put `text-grey-2 text-grey-4-hover` on it to match the other dashcard actions

 screenshot 2018-06-07 13 28 24

**Comment by r-kot** [comment]
@tlrobinson: The download icon has the same color as the title and the information icon
obraz
Should I change only the color of the download icon, or all colors in the header?

**Comment by mazameli** [comment]
Sorry for the delayed answer, @r-kot. I think we should change both the download icon as well as the information icon to have the classes that @tlrobinson mentioned, but we should leave the color of the card title alone.

**Comment by r-kot** [comment]
Hi, @mazameli. I did what you asked.

**Comment by sugengsuprayogi** [comment]
+1

**Comment by mazameli** [comment]
Thanks, @r-kot. I just tried this out and it looks ready to go. 

@metabase/core-developers Any other objections to merging this?

screen shot 2018-07-06 at 10 59 43 am

screen shot 2018-07-06 at 10 59 47 am

**Comment by camsaul** [comment]
WFM

**Comment by goodfella1408** [comment]
Thanks @r-kot for taking this forward and getting the changes merged.
One of the issues I have observed is that the downloads do not honour the column name changes done on metabase admin console

**Comment by arnoldagolli** [comment]
When will this be available in the live version, anyone knows?

**Comment by FleetOpsLiam** [comment]
@camsaul / @salsakran - would this have made its way into the Enterprise version? 1.0.3 ?

**Comment by erwan-boulard** [comment]
In which version will this be available ?

Thks

**Comment by flamber** [comment]
@erwan-boulard This has been available since version 0.30 if I remember correctly. Note that it's specific to Embedded dashboards, not Public dashboards (upvote #6687 by clicking :+1: on first post).
Please use the forum for questions and troubleshooting: https://discourse.metabase.com/
